test/e2e_federation/upgrades/generic.go

```diff
@@ -0,0 +1,65 @@
+/*
+Copyright 2017 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package upgrades
+
+import (
+	"fmt"
+
+	pkgruntime "k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/kubernetes/federation/pkg/crud"
+	crudutil "k8s.io/kubernetes/federation/pkg/crud/util"
+	fedframework "k8s.io/kubernetes/test/e2e_federation/framework"
+
+	. "github.com/onsi/ginkgo"
+)
+
+// GenericUpgradeTest validates that a federated resource remains
+// propagated before and after a controlplane upgrade
+type GenericUpgradeTest struct {
+	adapter crud.ResourceAdapter
+	crud    *crudutil.CRUDHelper
+	obj     pkgruntime.Object
+}
+
+// Setup creates a resource and validates its propagation to member clusters
+func (t *GenericUpgradeTest) Setup(f *fedframework.Framework) {
```

Comment:
While your logic regarding SetUp vs Setup may be sound, I'm just being consistent with the interface I'm implementing:

- https://github.com/kubernetes/kubernetes/blob/master/test/e2e_federation/upgrades/upgrade.go#L43

The federation upgrade interface is based on the kubernetes one:

 - https://github.com/kubernetes/kubernetes/blob/master/test/e2e/upgrades/upgrade.go#L46